File: 2019/Day 02/short.rb
def eval(xs, n = 12, v = 2)
  j, xs = -4, xs.clone
  xs[1], xs[2] = n, v
  while xs[j += 4] != 99
    op, a, b, c = xs[j, 4]
    proc = lambda(&(op.odd? ? :+ : :*))
    xs[c] = proc.call(xs[a], xs[b])
  end; xs.first
end

def solve(xs, t = 19690720)
  b = eval(xs, 0, 0)
  mj = eval(xs, 1, 0) - b
  v = (t -= b) % (n = t / mj)
  [100 * n, v].sum
end

xs = File.read('input.txt').split(',').map(&:to_i)
puts "Silver: #{eval(xs)}", "Gold: #{solve(xs)}"

File: 2019/Day 02/short.hs
{-#LANGUAGE QuasiQuotes #-}

import Data.Function (on)
import Control.Lens hiding (re)
import Data.Vector hiding (map, zip)
import Control.Lens.At (iix)
import Text.RE.TDFA.String
import Text.Printf

main :: IO()
main = do
  xs <- input
  printf "Silver: %d\n"
    $ eval xs [12, 2]
  printf "Gold: %d\n"
    $ values xs

values :: Vector Int -> Int
values xs = 100 * n + v where
  (n, v) = divMod t $
    eval xs [1, 0] - b
  t = 19690720 - b
  b = eval xs [0, 0]

eval :: Vector Int -> [Int] -> Int
eval arr nv = run 0 $arr // zip [1..] nv where
  run k xs = case zx of
    99 -> unsafeHead xs; 2 -> op (*); 1 -> op (+); where
      op fn = run (k + 4) (xs & iix c .~ on fn (xs!) a b)
      [zx, a, b, c] = toList $ slice k 4 xs

input :: IO (Vector Int)
input = fromList.regex <$> readFile "input.txt" where
  regex = map read.matches.(*=~ [re|@{%int}|])
